### PR TITLE
feat: Remove `Share` struct, add documentation, add `rc` constructor variants

### PR DESCRIPTION
**Pull request by appcypher**
## Summary

This PR adds and updates existing documentation, doc tests and makes other API clean ups.

- [x] Add more information to docs home page
- [x] Remove unused types and Share struct
- [x] Update READMEs
- [x] Add rc constructor variants

## Test plan (required)
-  Testing the code.
    
    ```bash
    scripts/wnfs.sh test
    ```

### CODE REVIEW

File: wnfs/src/lib.rs [hunk 1]
```diff
@@ -4,8 +4,122 @@
...
+//! }
+//! ```
+//!
+//! This example introduces a few new concepts. The first is the `HamtForest` which is a HAMT that can contain multiple file trees and implements the `PrivateForest` interface needed for persisting private file systems.
+//!
+//! The second is the `Name` (returned from `forest.empty_name()`) and `NameAccumulator` that lets us identify nodes in the filesystem, and are suitable for offspring proving.
+//!
+//! Finally, we have the random number generator, `rng`, that the library uses for generating new keys and other random values needed for the protocol.
+//!
+//! Check the [`wnfs/examples/`](wnfs-examples) folder for more examples.
```
Comment (matheus23) [inline wnfs/src/lib.rs:113]: Barney thumbs up Gif

**Review by matheus23 — COMMENTED** [review]
Big fan of most changes, just not a big fan of saving a bunch of characters with the `rc` name.
Wdyt about `new_rc`? Or perhaps an `rc` function with the signature `fn rc(self) -> Rc<Self>)`? So it's `&mut PrivateDirectory::new(...).rc()` (perhaps that's too easy to miss in examples though)?

**Comment by appcypher** [comment]
> Or perhaps an rc function with the signature fn rc(self) -> Rc<Self>? So it's &mut PrivateDirectory::new(...).rc() (perhaps that's too easy to miss in examples though)?

Thought about making an `rc` function; probably should go with that.
I know this change adds little but I think it simplifies the interface makes it less verbose. We almost always wrap these types in Rcs. They have little use outside of that so I think it only makes sense to make it part of the API.
Alsop it is in line with functions like `Box::pin` that wraps a Box in Pin. There are similar functions like this in the ecosystem.

**Comment by matheus23** [comment]
> Also, it is in line with functions like `Box::pin` that wraps a Box in Pin. There are similar functions like this in the ecosystem.

hmm I see. I'd argue `Box` is just inherently a simpler type (& `Box::new` thus a simpler function), since it's just a wrapper.
The constructor for `PrivateDirectory::new` takes multiple arguments and is more complex than a wrapper.

What do you think of calling the function `new_rc`?

Also. Perhaps worth trying this out: How crazy would it be if `new` simply always returned an `Rc`? Do we have code in rs-wnfs that would break? Maybe we can base our whole API around `Rc<PrivateDirectory>` and `&mut Rc<PrivateDirectory>`...

**Review by matheus23 — APPROVED** [review]
Addresses some of the #315

**Comment by appcypher** [comment]
@matheus23 Thanks for fixing 👍. Was going to get to it today.

**Pull request merged**